File: APPLICATION/models/cliente.py
"""
cliente.py — Modelo de Cliente con serialización binaria de longitud fija.

Estructura del registro (131 bytes por registro):
┌──────────┬──────────────────┬──────────┬──────────────┬──────────────────┬──────────┐
│  id      │  nombre          │  dni     │  telefono    │  email           │  estado  │
│  4 bytes │  50 bytes        │  11 bytes│  15 bytes    │  50 bytes        │  1 byte  │
└──────────┴──────────────────┴──────────┴──────────────┴──────────────────┴──────────┘
Total: 4 + 50 + 11 + 15 + 50 + 1 = 131 bytes

Offsets:
  id       → 0
  nombre   → 4
  dni      → 54
  telefono → 65
  email    → 80
  estado   → 130
"""
import struct
from dataclasses import dataclass
from typing import ClassVar


@dataclass
class Cliente:
    """
    Representa un cliente del minimarket.

    Serialización: struct big-endian "!I50s11s15s50sB"
    Tamaño fijo:   131 bytes por registro
    """

    # ! = big-endian
    # I   = unsigned int (4 bytes)  → id
    # 50s = char[50]    (50 bytes)  → nombre
    # 11s = char[11]    (11 bytes)  → dni
    # 15s = char[15]    (15 bytes)  → telefono
    # 50s = char[50]    (50 bytes)  → email
    # B   = unsigned char (1 byte)  → estado
    FORMAT: ClassVar[str] = "!I50s11s15s50sB"
    SIZE:   ClassVar[int] = struct.calcsize("!I50s11s15s50sB")  # 131

    id:       int = 0
    nombre:   str = ""
    dni:      str = ""
    telefono: str = ""
    email:    str = ""
    estado:   int = 1   # 1=activo, 0=eliminado

    def pack(self) -> bytes:
        """Serializa el cliente a 131 bytes de longitud fija."""
        return struct.pack(
            self.FORMAT,
            self.id,
            self.nombre.encode("utf-8")[:50].ljust(50, b"\x00"),
            self.dni.encode("utf-8")[:11].ljust(11, b"\x00"),
            self.telefono.encode("utf-8")[:15].ljust(15, b"\x00"),
            self.email.encode("utf-8")[:50].ljust(50, b"\x00"),
            self.estado,
        )

    @classmethod
    def unpack(cls, data: bytes) -> "Cliente":
        """Deserializa 131 bytes a un objeto Cliente."""
        if len(data) != cls.SIZE:
            raise ValueError(
                f"Cliente.unpack: se esperaban {cls.SIZE} bytes, "
                f"se recibieron {len(data)}"
            )
        raw = struct.unpack(cls.FORMAT, data)
        return cls(
            id=raw[0],
            nombre=raw[1].rstrip(b"\x00").decode("utf-8", errors="replace").strip(),
            dni=raw[2].rstrip(b"\x00").decode("utf-8", errors="replace").strip(),
            telefono=raw[3].rstrip(b"\x00").decode("utf-8", errors="replace").strip(),
            email=raw[4].rstrip(b"\x00").decode("utf-8", errors="replace").strip(),
            estado=raw[5],
        )

    def to_dict(self) -> dict:
        return {
            "id": self.id, "nombre": self.nombre, "dni": self.dni,
            "telefono": self.telefono, "email": self.email, "estado": self.estado,
        }

    def to_csv_row(self) -> str:
        return f"{self.id},{self.nombre},{self.dni},{self.telefono},{self.email},{self.estado}"

    def __repr__(self) -> str:
        return (
            f"Cliente(id={self.id}, nombre={self.nombre!r}, "
            f"dni={self.dni!r}, estado={self.estado})"
        )

File: APPLICATION/src/main/java/com/minimarket/models/Cliente.java
package com.minimarket.models;

import java.io.*;
import java.nio.charset.StandardCharsets;

/**
 * Modelo de Cliente con serialización binaria de longitud fija.
 *
 * Estructura del registro (131 bytes):
 *   id(4B) + nombre(50B) + dni(11B) + telefono(15B) + email(50B) + estado(1B)
 *
 * Compatible con Python: struct.pack("!I50s11s15s50sB", ...)
 */
public class Cliente implements BinaryRecord {

    public static final int SIZE = 131; // 4 + 50 + 11 + 15 + 50 + 1

    private int    id;
    private String nombre;
    private String dni;
    private String telefono;
    private String email;
    private int    estado;

    public Cliente() {
        this.nombre   = "";
        this.dni      = "";
        this.telefono = "";
        this.email    = "";
        this.estado   = 1;
    }

    public Cliente(int id, String nombre, String dni, String telefono, String email, int estado) {
        this.id       = id;
        this.nombre   = safe(nombre);
        this.dni      = safe(dni);
        this.telefono = safe(telefono);
        this.email    = safe(email);
        this.estado   = estado;
    }

    // ── Serialización ────────────────────────────────────────────────────────

    @Override
    public byte[] toBytes() {
        ByteArrayOutputStream bos = new ByteArrayOutputStream(SIZE);
        DataOutputStream dos = new DataOutputStream(bos);
        try {
            dos.writeInt(id);
            writeFixed(dos, nombre,   50);
            writeFixed(dos, dni,      11);
            writeFixed(dos, telefono, 15);
            writeFixed(dos, email,    50);
            dos.writeByte(estado);
        } catch (IOException e) {
            throw new RuntimeException("Error serializando Cliente", e);
        }
        return bos.toByteArray();
    }

    public static Cliente fromBytes(byte[] data) throws IOException {
        if (data.length != SIZE) {
            throw new IllegalArgumentException(
                "Cliente.fromBytes: se esperaban " + SIZE + " bytes, se recibieron " + data.length);
        }
        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(data));
        Cliente c = new Cliente();
        c.id       = dis.readInt();
        c.nombre   = readFixed(dis, 50);
        c.dni      = readFixed(dis, 11);
        c.telefono = readFixed(dis, 15);
        c.email    = readFixed(dis, 50);
        c.estado   = dis.readByte() & 0xFF;
        return c;
    }

    // ── Exportación ──────────────────────────────────────────────────────────

    public String toCsvRow() {
        return String.format("%d,%s,%s,%s,%s,%d",
            id, csv(nombre), csv(dni), csv(telefono), csv(email), estado);
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private static void writeFixed(DataOutputStream dos, String s, int len) throws IOException {
        byte[] src = safe(s).getBytes(StandardCharsets.UTF_8);
        byte[] dst = new byte[len];
        System.arraycopy(src, 0, dst, 0, Math.min(src.length, len));
        dos.write(dst);
    }

    private static String readFixed(DataInputStream dis, int len) throws IOException {
        byte[] bytes = new byte[len];
        dis.readFully(bytes);
        int end = 0;
        while (end < len && bytes[end] != 0) end++;
        return new String(bytes, 0, end, StandardCharsets.UTF_8).strip();
    }

    private static String safe(String s) { return s != null ? s : ""; }

    private static String csv(String s) {
        if (s == null) return "";
        if (s.contains(",") || s.contains("\"")) return "\"" + s.replace("\"", "\"\"") + "\"";
        return s;
    }

    // ── BinaryRecord ─────────────────────────────────────────────────────────

    @Override public int getId()           { return id; }
    @Override public void setId(int id)    { this.id = id; }
    @Override public int getEstado()       { return estado; }
    @Override public void setEstado(int e) { this.estado = e; }

    // ── Getters / Setters ────────────────────────────────────────────────────

    public String getNombre()              { return nombre; }
    public void   setNombre(String v)      { nombre = safe(v); }
    public String getDni()                 { return dni; }
    public void   setDni(String v)         { dni = safe(v); }
    public String getTelefono()            { return telefono; }
    public void   setTelefono(String v)    { telefono = safe(v); }
    public String getEmail()               { return email; }
    public void   setEmail(String v)       { email = safe(v); }

    @Override
    public String toString() {
        return String.format("Cliente(id=%d, nombre='%s', dni='%s', estado=%d)",
            id, nombre, dni, estado);
    }
}
